### DISCUSSION TITLE
🖐️ Reading files from both containers and host as an array of bytes, not UTF-8 string

### DISCUSSION THREAD

**Discussion by xsxz** [dagger/dagger#12428] [Help]
### What happened? What did you expect to happen?

I'm trying to extract UNIX utilities from pre-built docker images, like kubectl and helm, using python script. As it is popular approach to cut some time while reusing already existing artifacts, instead of building and storing tar image.
 
As far as I understand: dagger engine, being written on Go, in which strings are always implicitly encoded in UTF-8, reads all system files as byte arrays by default. While using 
```
host_file = await (client.host().file("./example.txt").contents())
print(host_file)
```
reading text files seems to be fine, but when I try to write contents of executable file 
```
host_file = await (client.host().file("./kubectl").contents())
with open(file="host_kubectl", mode="w") as f:
                    f.write(host_file)
```
it always results in corrupted and inflated file: `ELF 64-bit LSB executable, *unknown arch 0xbfef* (SYSV), no program header, corrupted section header size`. 
The main idea, why I'm trying to write file content directly to variable, instead of just copying files from container to host, and then to another container, is avoiding writing files to disk, so there are no cached or junk files, left after manipulations with them, forcing to cleanup every time after running pipeline.

Am I missing anything, like escaped characters from compiled file? Or am I just doing it wrong way? pls halp

**Comment by sipsma** [comment]
Re-opening because I think this is a good request, it would make sense to also support reading the raw bytes of files in addition to reading as a utf string.

The reason this hasn't come up very much previously is that so far the most common use cases don't want to read binaries directly; the typically end up just exporting them to the caller's host and/or publishing them as part of a container image.

I think the simplest route would be to just add an extra API like `ByteContents` that returns `[]byte` in Go and equivalent in all other SDK langs.

**Comment by Harry93** [comment]
Hi there, 

first of all I love the dagger.io from the very beginning. Great progress and nice features!

The past days I worked a lot with dagger and there is something, which bothers me and is kind of related to this issue. I implemented a function (maybe not following best practices but you should get the point), which zips the directory. The same can be done for tar and tar.gz.

```go
func Zip(ctx context.Context, client *dagger.Client, directory *dagger.Directory) *dagger.File {
	zipContainer := client.Container().
		From("alpine").
		WithExec([]string{"apk", "update"}).
		WithExec([]string{"apk", "add", "--no-cache", "zip"}).
		WithDirectory("/tmp/directory", directory).
		WithWorkdir("/tmp/directory").
		WithExec([]string{"zip", "-r", "/tmp/directory.zip", "."})

	return zipContainer.File("/tmp/directory.zip")
}
```

Since this takes some time to load the image, install everything (kind of tricky in regulated build environments behind proxies and firewalls), I implemented  the zip compression of these files within the go and in memory. So I implemented a zip function for the dagger.Directory, which worked perfectly fine. Unfortunately, when writing the resulting `var zipArchive []byte` array to dagger using `client.Directory().WithNewFile("xxx", dagger.ContainerWithNewFileOpts{Contents: string(zipArchive)})` the MarshalGQL converts only the valid utf-8 symbols and the written`\xff` within the query corrupts the final archive.

Long story short: When you guys think about supporting binary files, please consider writing binary files, too.

Learned quite a lot the past week.

**Comment by gerhard** [comment]
Both of these are sensible & relevant asks, will pick up as time & other priorities permit. Contributions welcome!

**Comment by gmile** [comment]
Just stumbled upon this while trying to read binary files using Elixir SDK. Today, on 0.15.1 below code will return an error:

<details>

```elixir
Mix.install [:dagger]

client = Dagger.connect!()

File.write!("/tmp/test-file.bin", << 1, 2, 3 >>) # <- binary contents

{:ok, contents} =
  client
  |> Dagger.Client.host()
  |> Dagger.Host.file("/tmp/test-file.bin")
  |> Dagger.File.contents()

{:ok, stdout} =
  client
  |> Dagger.Client.container([])
  |> Dagger.Container.from("alpine:3.21")
  |> Dagger.Container.with_new_file("file", contents)
  |> Dagger.Container.with_exec(~w"cat file")
  |> Dagger.Container.stdout()

IO.puts(stdout)

Dagger.close(client)
```

Error from the log:

```elixir
{:error,
 %Dagger.Core.GraphQL.Response{
   data: nil,
   extensions: nil,
   errors: [
     %Dagger.Core.GraphQL.Response.Error{
       message: "Unexpected <Invalid>",
       path: nil,
       locations: [%{"column" => 67, "line" => 1}],
       extensions: %{"code" => "GRAPHQL_PARSE_FAILED"}
     }
   ]
 }}
```

If I replace `<< 1, 2, 3 >>` with `"Hello, world!"` everything is working.

</details>

My use-case to read binary is to generate sha256 hash of its contents using stdlib of the runtime I'm in (Erlang VM in my case), and then do some other stuff to the hash. Not sure if this is idiomatic way to do so, since I could install and run `shasum -a 256` in the container on the file, and then extract it. It just kind of felt a bit more natural to do so in the runtime for my case, where I actually need to generate shasums of multiple files, then process the result and save it.

**Comment by mkaramuk** [comment]
> Re-opening because I think this is a good request, it would make sense to also support reading the raw bytes of files in addition to reading as a utf string.
> 
> The reason this hasn't come up very much previously is that so far the most common use cases don't want to read binaries directly; the typically end up just exporting them to the caller's host and/or publishing them as part of a container image.
> 
> I think the simplest route would be to just add an extra API like `ByteContents` that returns `[]byte` in Go and equivalent in all other SDK langs.

I was trying to read the compiled binary file and attach it into the GitHub release and seems like this approach is the only way to do that, at least for now.
